cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1001,47 +1004,77 @@ struct CastFunctor<TimestampType, StringType> {
 // Binary to String
 //
 
+#if defined(_MSC_VER)
+// Silence warning: """'visitor': unreferenced local variable"""
+#pragma warning(push)
+#pragma warning(disable : 4101)
+#endif
+
 template <typename I>
 struct CastFunctor<StringType, I,
                    typename std::enable_if<std::is_same<BinaryType, I>::value>::type> {
   void operator()(FunctionContext* ctx, const CastOptions& options,
                   const ArrayData& input, ArrayData* output) {
     BinaryArray binary(input.Copy());
 
-    if (options.allow_invalid_utf8) {
-      ZeroCopyData(input, output);
-      return;
+    if (!options.allow_invalid_utf8) {
+      util::InitializeUTF8();
+
+      ArrayDataVisitor<I> visitor;
+      Status st = visitor.Visit(input, this);
+      if (!st.ok()) {
+        ctx->SetStatus(st);
+        return;
+      }
     }
+    ZeroCopyData(input, output);
+  }
 
-    util::InitializeUTF8();
+  Status VisitNull() { return Status::OK(); }
 
-    if (binary.null_count() != 0) {
-      for (int64_t i = 0; i < input.length; i++) {
-        if (binary.IsNull(i)) {
-          continue;
-        }
+  Status VisitValue(util::string_view str) {
+    if (ARROW_PREDICT_FALSE(!arrow::util::ValidateUTF8(str))) {
+      return Status::Invalid("Invalid UTF8 payload");
+    }
+    return Status::OK();
+  }
+};
 
-        const auto str = binary.GetView(i);
-        if (ARROW_PREDICT_FALSE(!arrow::util::ValidateUTF8(str))) {
-          ctx->SetStatus(Status::Invalid("Invalid UTF8 payload"));
-          return;
-        }
-      }
+template <typename I>
+struct CastFunctor<
```

Comment:
You can forgo enable_if here, see cast(timestamp -> string)
```suggestion
struct CastFunctor<LargeStringType, LargeBinaryType>
```

Reply:
Oh, you're right. I wonder why the file used this coding style.

Reply:
Also, I can refactor the common code here.